### ISSUE TITLE
No localized low-zoom labels when system language is Simplified Chinese

### ISSUE THREAD

**Issue by 1ec5** [mapbox/mapbox-gl-native#12164]
When the system language is set to Simplified Chinese, labels of countries, states and provinces, and oceans and seas disappear instead of being localized into Simplified Chinese. This is because `+[MGLVectorTileSource preferredMapboxStreetsLanguageForPreferences:]` assumes all Streets source layers offer the same languages. In fact, [this documentation](https://www.mapbox.com/vector-tiles/mapbox-streets-v7/#overview) says (in a rather confusing way) that `name_zh-Hans` is unavailable for the `country_label`, `state_label`, and `marine_label` layers.

The language fallback logic being added for #11867 could address this issue, but a more targeted fix would involve special-casing these three source layers.

/ref #11867
/cc @friedbunny @nickidlugash @langsmith

**Comment by lilykaiser** [comment]
@lloydsheng do you have time to investigate this? cc @chriswu42

**Comment by lloydsheng** [comment]
@lilykaiser Happy to help investigate this issue.  cc @suntony

**Comment by lloydsheng** [comment]
@1ec5 I checked [the tile data](https://github.com/mapbox/mapbox-gl-native/files/2188246/mapbox.mapbox-terrain-v2.mapbox.mapbox-streets-v7.1.1.0.vector.pbf.zip), These layers did have no `name_zh-Hans` field.

| country_label| marine_label   | state_label  |
| ------------- |:-------------:| -------------:|
|  ![image](https://user-images.githubusercontent.com/503040/42629047-a818bd30-8604-11e8-8ffd-8a34fd2827b5.png)     | ![image](https://user-images.githubusercontent.com/503040/42629186-26be5a00-8605-11e8-8fc2-b31eca8ee3d3.png) |![image](https://user-images.githubusercontent.com/503040/42629222-4159fd42-8605-11e8-9707-fe92ede731ca.png) |

My initial thought was that can we add the `name_zh-Hans` field to those layers from the data side? If can't, I have 2 options to fix it from SDK side.

1.   Check the layer when setting layer's text, if it is one of  `country_label`, `state_label`,`marine_label`, then check the `localizedText` expression, fallback to `name_zh` if it is `name_zh-Han`.

2. As we have the expression now, we could add a logic with the expression to check all the name field,  fallback to `name_zh` if it is `name_zh-Han` and the value of it is empty.

@julianrex @1ec5 Glad to hear your thougts about this.

**Comment by 1ec5** [comment]
> My initial thought was that can we add the `name_zh-Hans` field to those layers from the data side?

Unfortunately, GL doesn’t currently support adding fields to vector tiles at runtime. (This is a concept sometimes called “data join”.) Better to alias one field with another in the expressions instead.

> Check the layer when setting layer's text, if it is one of  `country_label`, `state_label`,`marine_label`, then check the `localizedText` expression, fallback to `name_zh` if it is `name_zh-Han`.

The navigation SDK has [similar logic](https://github.com/mapbox/mapbox-navigation-ios/blob/e2b3f4a53742ce0c092de7ba226ecd535e7221df/MapboxNavigation/NavigationMapView.swift#L971-L977) that determines the locale to pass into `-[NSExpression mgl_expressionLocalizedIntoLocale:]` based on the layer’s source layer identifier.

The special case would have to live in `-[MGLStyle localizeLabelsIntoLocale:]`, because `-[NSExpression mgl_expressionLocalizedIntoLocale:]` doesn’t know the current source name’s locale (and giving it that context would make little sense and would break backwards compatibility). If we put the special case in `-[MGLStyle localizeLabelsIntoLocale:]`, then the caller of `-[NSExpression mgl_expressionLocalizedIntoLocale:]` is responsible for making sure the locale is appropriate for the current layer.

> As we have the expression now, we could add a logic with the expression to check all the name field,  fallback to `name_zh` if it is `name_zh-Han` and the value of it is empty.

The fallback could be implemented by wrapping key path expressions in `mgl_coalesce()` functions. This is the more robust approach. We need to take this approach anyways for mapbox/mapbox-gl-native#11867, since all the `name_` fields in Streets source v8 are optional. The tricky part is that localizing the style a second time would needlessly complicate the expression. Localizing an already localized expression with the same locale could turn:

```swift
NSExpression(format: "mgl_coalesce({name_zh-Hans, name_zh})")
```

into:

```swift
NSExpression(format: """
    mgl_coalesce({
        mgl_coalesce({name_zh-Hans, name_zh}),
        mgl_coalesce({name_zh-Hans, name_zh})
    })
    """)
```

With a different locale the second time, it would become:

```swift
NSExpression(format: """
    mgl_coalesce({
        name_en,
        name_en
    })
    """)
```

If we could positively identify a coalescing expression as the result of an existing localization operation, then we might be able to replace the expression instead of more deeply nesting it, somewhere around here:

https://github.com/mapbox/mapbox-gl-native/blob/192c61146847eda5b62cf78b0987aeb59a17ef0b/platform/darwin/src/NSExpression+MGLAdditions.mm#L1519-L1530

**Comment by lloydsheng** [comment]
@1ec5 Thanks for your feedbacks 👍. I've pushed some changes to fix this issue https://github.com/mapbox/mapbox-gl-native/pull/12387. I think it also made the SDK compatible with name fields of Mapbox Streets v8.

**Comment by 1ec5** [comment]
> If we could positively identify a coalescing expression as the result of an existing localization operation, then we might be able to replace the expression instead of more deeply nesting it

We could append a dummy key path expression to the end of the coalescing array, something like `com.mapbox.expressions.localized`. Right around here, before localizing the arguments, we could check whether the dummy key path appears in the coalescing expression:

https://github.com/mapbox/mapbox-gl-native/blob/cb714d57c5c5ad181aaf5e1690221da8d965682b/platform/darwin/src/NSExpression%2BMGLAdditions.mm#L1414

If it does, then we can replace the expression with a new localized one instead of relocalizing the existing expression (which would cause it to become more deeply nested):

```objc
NSExpression *localizedCoalescingExpressionMarker = [NSExpression expressionForKeyPath:@"com.mapbox.expressions.localized"];

if ([self.function isEqualToString:@"mgl_coalesce:"]) {
    // Check if the coalescing expression was the result of previously localizing an expression.
    NSArray *coalescedExpressions = (NSArray *)self.arguments.firstObject;
    if ([coalescedExpressions isKindOfClass:[NSArray array]] &&
        [coalescedExpressions.lastObject isEqual:localizedCoalescingExpressionMarker]) {
        // If so, replace the expression instead of relocalizing it.
        return [[NSExpression expressionForKeyPath:@"name"]
                mgl_expressionLocalizedIntoLocale:locale];
    }
}

NSArray *localizedArguments = MGLLocalizedCollection(arguments, locale);
// …
```

**Comment by 1ec5** [comment]
Fixed in #12387.
